src/parser/addressRepresentationConverters.ts

```diff
@@ -31,54 +31,66 @@ export const cellAddressFromString = (sheetMapping: SheetMappingFn, stringAddres
 
   const col = columnLabelToIndex(result[6])
 
-  const sheet = extractSheetNumber(result, sheetMapping)
+  let sheet = extractSheetNumber(result, sheetMapping)
   if (sheet === undefined) {
     return undefined
   }
 
+  if(sheet === null) {
+    sheet = undefined
+  }
+
   const row = Number(result[8]) - 1
   if (result[5] === '$' && result[7] === '$') {
-    return CellAddress.absolute(sheet, col, row)
+    return CellAddress.absolute(col, row, sheet)
   } else if (result[5] === '$') {
-    return CellAddress.absoluteCol(sheet, col, row - baseAddress.row)
+    return CellAddress.absoluteCol( col, row - baseAddress.row, sheet)
   } else if (result[7] === '$') {
-    return CellAddress.absoluteRow(sheet, col - baseAddress.col, row)
+    return CellAddress.absoluteRow(col - baseAddress.col, row, sheet)
   } else {
-    return CellAddress.relative(sheet, col - baseAddress.col, row - baseAddress.row)
+    return CellAddress.relative(row - baseAddress.row, col - baseAddress.col, sheet)
   }
 }
 
 export const columnAddressFromString = (sheetMapping: SheetMappingFn, stringAddress: string, baseAddress: SimpleCellAddress): Maybe<ColumnAddress> => {
   const result = columnRegex.exec(stringAddress)!
 
-  const sheet = extractSheetNumber(result, sheetMapping)
+  let sheet = extractSheetNumber(result, sheetMapping)
   if (sheet === undefined) {
     return undefined
   }
 
+  if(sheet === null) {
+    sheet = undefined
+  }
+
   const col = columnLabelToIndex(result[6])
 
   if (result[5] === '$') {
-    return ColumnAddress.absolute(sheet, col)
+    return ColumnAddress.absolute(col, sheet)
   } else {
-    return ColumnAddress.relative(sheet, col - baseAddress.col)
+    return ColumnAddress.relative(col - baseAddress.col, sheet)
   }
 }
 
 export const rowAddressFromString = (sheetMapping: SheetMappingFn, stringAddress: string, baseAddress: SimpleCellAddress): Maybe<RowAddress> => {
   const result = rowRegex.exec(stringAddress)!
 
-  const sheet = extractSheetNumber(result, sheetMapping)
+  let sheet = extractSheetNumber(result, sheetMapping)
   if (sheet === undefined) {
     return undefined
   }
 
+  if(sheet === null) {
+    sheet = undefined
+  }
+
```

Comment:
When `sheet` is `undefined` we exit early, but then we assign `undefined` and process it. Shouldn't we exit early for both `null` and `undefined`?

Reply:
Yeah, this part is a mess. Believe me, the logic should be as it is (there are two distinct failure modes here), and it is pretty annoying. I really don't see any clean way of making it less ugly, though.